renderdoc/driver/vulkan/vk_core.h

```diff
@@ -735,59 +745,121 @@ class WrappedVulkan : public IFrameCapturer
   };
   rdcarray<ActionUse> m_ActionUses;
 
-  enum PartialReplayIndex
+  // during active replay, command buffers may be partially-submitted if the selected event occurs
+  // within the range of the command buffer. If secondary command buffers are used and the selected
+  // event occurs within the range of a secondary, both the secondary and the command buffer that
+  // executed it are considered partially-submitted. If VK_EXT_nested_command_buffer is enabled,
+  // there may be up to N partially-submitted command buffers, where N is the maximum command
+  // nesting depth used in the capture.
+  struct PartialReplayData
   {
-    Primary,
-    Secondary,
-    ePartialNum
-  };
+    PartialReplayData() { Reset(); }
+    void Reset() { partialStack.clear(); }
+
+    rdcarray<CommandBufferNode> partialStack;
+    rdcarray<CommandBufferNode> commandTree;
+    rdcflatmap<ResourceId, rdcarray<CommandBufferNode>> submitLookup;
+  } m_Partial;
 
-  struct Submission
+  rdcflatmap<ResourceId, CommandBufferNode> m_CommandNodes;
+
+  bool IsCommandBufferPartial(ResourceId cmdId)
   {
-    Submission(uint32_t eid) : baseEvent(eid), rebased(false) {}
-    uint32_t baseEvent = 0;
-    bool rebased = false;
-  };
+    for(const CommandBufferNode &cmdNode : m_Partial.partialStack)
+    {
+      if(cmdNode.cmdId == cmdId)
+        return true;
+    }
 
-  // by definition, when replaying we must have N completely submitted command buffers, and at most
-  // two partially-submitted command buffers. One primary, that we're part-way through, and then
-  // if we're part-way through a vkCmdExecuteCommandBuffers inside that primary then there's one
-  // secondary.
-  struct PartialReplayData
+    return false;
+  }
+
+  bool IsCommandBufferPartialPrimary(ResourceId cmdId)
   {
-    PartialReplayData() { Reset(); }
-    void Reset()
+    if(m_Partial.partialStack.empty())
+      return false;
+
+    return (m_Partial.partialStack.front().cmdId == cmdId) &&
+           (m_BakedCmdBufferInfo[cmdId].level == VK_COMMAND_BUFFER_LEVEL_PRIMARY);
+  }
+
+  void SetPartialStack(const CommandBufferNode &targetNode, uint32_t curEvent)
```

Comment:
As a minor and admittedly quite arbitrary point, could you move these functions `SetPartialStack`, `BuildPartialStackUpToTarget`, and `BuildPartialStack` to the source file? Usually for in-line functions I limit to only accessor-type things - a lookup and a check or something like that. The other functions here are fine to leave in the header.